src/boot/lib/ast.ml

```diff
@@ -478,6 +480,139 @@ let ty_info = function
   | TyApp (fi, _, _) ->
       fi
 
+let const_has_side_effect = function
+  | CBool _
+  | CInt _
+  | Caddi _
+  | Csubi _
+  | Cmuli _
+  | Cdivi _
+  | Cmodi _
+  | Cnegi
+  | Clti _
+  | Cleqi _
+  | Cgti _
+  | Cgeqi _
+  | Ceqi _
+  | Cneqi _
+  | Cslli _
+  | Csrli _
+  | Csrai _
+  | Carity ->
+      false
+  (* MCore intrinsics: Floating-point numbers *)
+  | CFloat _
+  | Caddf _
+  | Csubf _
+  | Cmulf _
+  | Cdivf _
+  | Cnegf
+  | Cltf _
+  | Cleqf _
+  | Cgtf _
+  | Cgeqf _
+  | Ceqf _
+  | Cneqf _
+  | Cfloorfi
+  | Cceilfi
+  | Croundfi
+  | Cint2float
+  | Cstring2float ->
+      false
+  (* MCore intrinsics: Characters *)
+  | CChar _ | Ceqc _ | Cchar2int | Cint2char ->
+      false
+  (* MCore intrinsic: sequences *)
+  | Ccreate _
+  | Clength
+  | Cconcat _
+  | Cget _
+  | Cset _
+  | Ccons _
+  | Csnoc _
+  | CsplitAt _
+  | Creverse
+  | Csubsequence _ ->
+      false
+  (* MCore intrinsics: Random numbers *)
+  | CrandIntU _ ->
+      true
+  | CrandSetSeed ->
+      true
+  (* MCore intrinsics: Time *)
+  | CwallTimeMs ->
+      true
+  | CsleepMs ->
+      true
+  (* MCore intrinsics: Debug and I/O *)
+  | Cprint
+  | Cdprint
+  | CreadLine
+  | CreadBytesAsString
+  | CreadFile
+  | CwriteFile _
+  | CfileExists
+  | CdeleteFile
+  | Cerror
+  | Cexit ->
+      true
+  (* MCore intrinsics: Symbols *)
+  | CSymb _ | Cgensym | Ceqsym _ | Csym2hash ->
+      true
```

Comment:
I don't think `eqsym` and `sym2hash` has side-effects. At least not in the current implementation.

Reply:
Might be correct now, but since the underlying handling of symbols has side effects, I would like to be conservative right now and just say that things that are guaranteed to be side effect free in the future are marked as such.